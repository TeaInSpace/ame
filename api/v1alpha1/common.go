package v1alpha1

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	runtime "k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/schema"
)

func GenOwnRef(obj metav1.ObjectMeta, gvks []schema.GroupVersionKind) metav1.OwnerReference {
	// TODO: Deal with case of len(gvks) == 0 and len(gvks)>1
	return metav1.OwnerReference{
		APIVersion: gvks[0].GroupVersion().String(), // TODO: is there a better method for getting the APIVersion string?
		Kind:       gvks[0].Kind,
		Name:       obj.GetName(),
		UID:        obj.GetUID(),
	}
}

func GenGvks(scheme *runtime.Scheme, obj runtime.Object) ([]schema.GroupVersionKind, error) {
	gvks, _, err := scheme.ObjectKinds(obj)
	if err != nil {
		return nil, err
	}

	return gvks, nil
}
