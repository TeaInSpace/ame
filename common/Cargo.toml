[package]
name = "common"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
controller = { version = "0.1.0", path = "../controller" }
either = "1.8.0"
k8s-openapi = "0.16.0"
kube = "0.76.0"
tokio = "1.23.0"
