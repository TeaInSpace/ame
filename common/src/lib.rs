use std::time::Duration;

use controller::{Task, Workflow};
use either::Either;
use k8s_openapi::api::core::v1::{LoadBalancerStatus, Service, ServiceSpec, ServiceStatus};
use kube::{
    api::{DeleteParams, ListParams},
    Api, Client,
};

pub async fn find_service_endpoint(
    namespace: &str,
    service_name: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let client = Client::try_default().await?;
    let services = Api::<Service>::namespaced(client.clone(), namespace);
    let service = services.get(service_name).await?;

    let Service { spec: Some(ServiceSpec{
       ports: Some(ports),
       ..
    }), status: Some(ServiceStatus{

        load_balancer: Some(LoadBalancerStatus{
            ingress: Some(ingress)
        }),
        ..
    }),
    ..} = service else {
        return Err(format!("failed to extract service ips and ports: {service:#?}"))?; 
    };

    if ingress.len() != 1 {
        return Err(format!(
            "expected a ingress but got {} for {} {:#?}",
            ingress.len(),
            service_name,
            ingress
        ))?;
    }

    if ports.len() != 1 {
        return Err(format!(
            "expected a port but got {} for {} {:#?}",
            ports.len(),
            service_name,
            ports
        ))?;
    }

    Ok(format!(
        "http://{}:{}",
        ingress[0].ip.as_ref().unwrap(),
        ports[0].port
    ))
}

/// Prepare a cluster for tests, under the assumptions that the `just setup_cluster` recipe has been run successfully.
/// This implies that all required custom resource definitions are installed in the cluster.
/// This function will generate clients and clear all Task and `Workflow` objects in the cluster.
pub async fn setup_cluster(
    namespace: &str,
) -> Result<(Api<controller::Task>, Api<controller::Workflow>), Box<dyn std::error::Error>> {
    let client = Client::try_default().await?;
    let tasks = Api::<Task>::namespaced(client.clone(), namespace);
    let workflows = Api::<Workflow>::namespaced(client.clone(), namespace);

    let dp = DeleteParams::default();
    let lp = ListParams::default();

    match tasks.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            while !tasks.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    match workflows.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            while !workflows.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    Ok((tasks, workflows))
}
