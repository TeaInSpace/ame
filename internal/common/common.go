package common

import (
	"context"
	"fmt"
	"time"

	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/fields"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/apimachinery/pkg/watch"
	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
	"teainspace.com/ame/api/v1alpha1"
	"teainspace.com/ame/internal/workflows"
)

type AmeKubeClient[obj any] interface {
	Get(context.Context, string, metav1.GetOptions) (obj, error)
	Update(context.Context, obj, metav1.UpdateOptions) (obj, error)
	Create(context.Context, obj, metav1.CreateOptions) (obj, error)
	Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
	Delete(context.Context, string, metav1.DeleteOptions) error
	DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error
}

type AmeGenClient[obj any] struct {
	Cli AmeKubeClient[obj]
}

func NewAmeGenClient[obj any](cli AmeKubeClient[obj]) AmeGenClient[obj] {
	return AmeGenClient[obj]{
		Cli: cli,
	}
}

func (c AmeGenClient[obj]) WaitForCondWithTimout(ctx context.Context, name string, verify func(obj) bool, timeout time.Duration) error {
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	return c.WaitForCond(ctx, name, verify)
}

func (c AmeGenClient[obj]) WaitForCond(ctx context.Context, name string, verify func(obj) bool) error {
	objChan, err := c.WatchObj(ctx, name)
	if err != nil {
		return err
	}

	for {
		select {
		case <-ctx.Done():
			return fmt.Errorf("from WaitForCondition, failed to match condition within timeout")
		case o := <-objChan:
			if verify(o) {
				return nil
			}
		}
	}
}

func (c AmeGenClient[obj]) IsCondition(ctx context.Context, name string, verify func(obj) bool) (bool, error) {
	kubeObj, err := c.Cli.Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return false, err
	}

	return verify(kubeObj), nil
}

func (c AmeGenClient[obj]) WatchObj(ctx context.Context, name string) (<-chan obj, error) {
	opts, err := GenNameListOpts(name)
	if err != nil {
		return nil, err
	}

	return c.Watch(ctx, opts)
}

func (c AmeGenClient[obj]) Watch(ctx context.Context, opts metav1.ListOptions) (<-chan obj, error) {
	watcher, err := c.Cli.Watch(ctx, opts)
	if err != nil {
		return nil, err
	}

	objChan := make(chan obj)
	results := watcher.ResultChan()

	go func() {
		defer close(objChan)

		for {
			select {
			case <-ctx.Done():
				return
			case e, ok := <-results:
				if !ok {
					return
				}

				objInstans, ok := e.Object.(obj)
				// TODO: How will we handle if ok=false
				if ok {
					objChan <- objInstans
				}
			}
		}
	}()

	return objChan, nil
}

func (c AmeGenClient[obj]) Create(ctx context.Context, o obj) (obj, error) {
	return c.Cli.Create(ctx, o, metav1.CreateOptions{})
}

func (c AmeGenClient[obj]) Get(ctx context.Context, name string) (obj, error) {
	return c.Cli.Get(ctx, name, metav1.GetOptions{})
}

func (c AmeGenClient[obj]) DeleteCollection(ctx context.Context, delOpts metav1.DeleteOptions, listOpts metav1.ListOptions) error {
	return c.Cli.DeleteCollection(ctx, delOpts, listOpts)
}

func WaitForPodPhase(ctx context.Context, pods clientv1.PodInterface, p *v1.Pod, targetPhase v1.PodPhase, timeout time.Duration) error {
	ameGenCli := NewAmeGenClient[*v1.Pod](pods)
	return ameGenCli.WaitForCondWithTimout(ctx, p.GetName(), func(p *v1.Pod) bool {
		return p.Status.Phase == targetPhase
	}, timeout)
}

func GenTaskPhaseValidator(phase v1alpha1.TaskPhase) func(*v1alpha1.Task) bool {
	return func(t *v1alpha1.Task) bool {
		return t.Status.Phase == phase
	}
}

func IsTaskRunning(t *v1alpha1.Task) bool {
	return t.Status.Phase == v1alpha1.TaskRunning
}

func GenNameListOpts(name string) (metav1.ListOptions, error) {
	selector, err := genNameSelector(name)
	if err != nil {
		return metav1.ListOptions{}, err
	}

	return metav1.ListOptions{
		FieldSelector: selector.String(),
	}, nil
}

func genNameSelector(name string) (fields.Selector, error) {
	return fields.ParseSelector("metadata.name=" + name)
}

func GenTaskPodSelector(name string) (labels.Selector, error) {
	return labels.Parse(fmt.Sprintf("%s=%s", workflows.TaskLabelKey, name))
}
