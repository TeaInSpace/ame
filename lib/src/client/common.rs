use futures::stream::StreamExt;
use http_body::Body;
use hyper::body::Bytes;
use tokio_stream::Stream;
use tonic::{codegen::StdError, Request};

use crate::{
    grpc::{ame_service_client::AmeServiceClient, ProjectSourceId, ProjectSourceStatus},
    Result,
};

#[derive(Clone)]
pub struct AmeCtrl<C> {
    pub client: AmeServiceClient<C>,
}

impl<'a, C> AmeCtrl<C>
where
    C: tonic::client::GrpcService<tonic::body::BoxBody> + 'a,
    C::Error: Into<StdError>,
    C::ResponseBody: Body<Data = Bytes> + Send + 'static,
    C: Clone,
    <C::ResponseBody as Body>::Error: Into<StdError> + Send,
{
    pub fn new(client: AmeServiceClient<C>) -> AmeCtrl<C> {
        AmeCtrl { client }
    }

    pub async fn watch_project_src(
        self,
        id: ProjectSourceId,
    ) -> Result<impl Stream<Item = ProjectSourceStatus> + 'a> {
        let initial_stream = self
            .client
            .clone()
            .watch_project_src(Request::new(id.clone()))
            .await?
            .into_inner();

        //TODO: ensure this does not turn into an unblocked loop when there is no server available.
        let res = Box::pin(async_stream::stream! {
            let mut stream = initial_stream;
            loop {
                    while let Some(Ok(value)) = stream.next().await {
                        yield value;
                    }

                    stream = self
                .client.clone()
                .watch_project_src(Request::new(id.clone()))
                .await
                .unwrap()
                .into_inner();

                }

        });

        Ok(res)
    }

    pub async fn delete_project_src(mut self, id: ProjectSourceId) -> Result<()> {
        self.client.delete_project_src(Request::new(id)).await?;
        Ok(())
    }
}
