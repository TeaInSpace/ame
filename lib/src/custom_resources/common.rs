use crate::{
    custom_resources::{argo::Workflow, data_set::DataSet, new_task::Task, project::Project},
    error::AmeError,
};
use either::Either;
use k8s_openapi::{
    api::core::v1::{LoadBalancerStatus, Service, ServiceSpec, ServiceStatus},
    apimachinery::pkg::apis::meta::v1::OwnerReference,
};
use kube::{
    api::{DeleteParams, ListParams, PatchParams},
    Api, Client, ResourceExt,
};

use std::time::Duration;

use super::project_source::ProjectSource;

pub async fn find_ame_endpoint(
    namespace: &str,
    service_name: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let client = Client::try_default().await?;
    let services = Api::<Service>::namespaced(client.clone(), namespace);
    let service = services.get(service_name).await?;

    let Service {
        spec: Some(ServiceSpec {
            ports: Some(ports), ..
        }),
        ..
    } = service
    else {
        return Err(format!(
            "failed to extract service ips and ports: {service:#?}"
        ))?;
    };

    let port = ports
        .iter()
        .find(|p| p.name.clone().unwrap_or("".to_string()) == "https");

    if let Some(port) = port {
        Ok(format!("https://ame.local:{}", port.port))
    } else {
        Err("failed to find a port".to_string())?
    }
}

pub async fn find_service_endpoint(
    namespace: &str,
    service_name: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let client = Client::try_default().await?;
    let services = Api::<Service>::namespaced(client.clone(), namespace);
    let service = services.get(service_name).await?;

    let Service {
        spec: Some(ServiceSpec {
            ports: Some(ports), ..
        }),
        status:
            Some(ServiceStatus {
                load_balancer:
                    Some(LoadBalancerStatus {
                        ingress: Some(ingress),
                    }),
                ..
            }),
        ..
    } = service
    else {
        return Err(format!(
            "failed to extract service ips and ports: {service:#?}"
        ))?;
    };

    if ingress.len() != 1 {
        return Err(format!(
            "expected a ingress but got {} for {} {:#?}",
            ingress.len(),
            service_name,
            ingress
        ))?;
    }

    if ports.len() != 1 {
        return Err(format!(
            "expected a port but got {} for {} {:#?}",
            ports.len(),
            service_name,
            ports
        ))?;
    }

    let Some(ip) = ingress[0].ip.as_ref() else {
        return Err("could not find IP from ingress".to_string().into());
    };

    Ok(format!("http://{}:{}", ip, ports[0].port))
}

pub fn parent_project(owner_references: Vec<OwnerReference>) -> crate::Result<String> {
    let projects: Vec<OwnerReference> = owner_references
        .into_iter()
        .filter(|o| kind_is_project(&o.kind))
        .collect();

    if projects.len() != 1 {
        return Err(AmeError::MissingProject(projects.len()));
    }

    Ok(projects[0].name.clone())
}

pub fn kind_is_project(kind: &str) -> bool {
    kind == "Project"
}

/// Prepare a cluster for tests, under the assumptions that the `just setup_cluster` recipe has been run successfully.
/// This implies that all required custom resource definitions are installed in the cluster.
/// This function will generate clients and clear all Task and `Workflow` objects in the cluster.
pub async fn setup_cluster(
    namespace: &str,
) -> Result<(Api<Task>, Api<Workflow>), Box<dyn std::error::Error>> {
    let client = Client::try_default().await?;
    let tasks = Api::<Task>::namespaced(client.clone(), namespace);
    let workflows = Api::<Workflow>::namespaced(client.clone(), namespace);
    let project_srcs = Api::<ProjectSource>::namespaced(client.clone(), namespace);
    let projects = Api::<Project>::namespaced(client.clone(), namespace);
    let data_sets = Api::<DataSet>::namespaced(client.clone(), namespace);

    let dp = DeleteParams::default();
    let lp = ListParams::default();
    let patch_params = PatchParams::apply("AME_TEST").force();

    for mut project in projects.list(&lp).await?.into_iter() {
        project.spec.deletion_approved = true;
        project.metadata.managed_fields = None;

        let res = projects
            .patch(
                &project.name_any(),
                &patch_params,
                &kube::api::Patch::Apply(project),
            )
            .await;

        match res {
            Err(kube::Error::Api(e)) => {
                if e.code == 409 {
                    return Ok((tasks, workflows));
                }
            }
            Err(e) => return Err(Box::new(e)),
            _ => (),
        }
    }

    match project_srcs.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            while !project_srcs.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };
    for mut task in tasks.list(&lp).await?.into_iter() {
        task.spec.deletion_approved = true;
        task.metadata.managed_fields = None;

        tasks
            .patch(
                &task.name_any(),
                &patch_params,
                &kube::api::Patch::Apply(task),
            )
            .await?;
    }

    match projects.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            for mut project in projects.list(&lp).await?.into_iter() {
                project.spec.deletion_approved = true;
                project.metadata.managed_fields = None;

                let _ = projects
                    .patch(
                        &project.name_any(),
                        &patch_params,
                        &kube::api::Patch::Apply(project),
                    )
                    .await;
            }
            while !projects.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    match data_sets.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            for mut data_set in data_sets.list(&lp).await?.into_iter() {
                data_set.spec.deletion_approved = true;

                data_sets
                    .patch(
                        &data_set.name_any(),
                        &PatchParams::default(),
                        &kube::api::Patch::Merge(data_set),
                    )
                    .await?;
            }

            while !data_sets.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    match tasks.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            for mut task in tasks.list(&lp).await?.into_iter() {
                task.spec.deletion_approved = true;
                task.metadata.managed_fields = None;

                tasks
                    .patch(
                        &task.name_any(),
                        &patch_params,
                        &kube::api::Patch::Apply(task),
                    )
                    .await?;
            }

            while !tasks.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    match workflows.delete_collection(&dp, &lp).await? {
        Either::Left(_) => {
            while !workflows.list(&lp).await?.items.is_empty() {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }

        Either::Right(status) => {
            println!("Deleted collection of tasks: {status:?}");
        }
    };

    Ok((tasks, workflows))
}

// This is an access token for https://github.com/TeaInSpace/ame-test-private intended for
// testing that AME can pull in projects from private repositories.
pub fn private_repo_gh_pat() -> Result<String, Box<dyn std::error::Error>> {
    Ok(std::env::var("AME_TEST_GH_TOKEN")?)
}
