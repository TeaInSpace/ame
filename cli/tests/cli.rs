use assert_cmd::prelude::*;
use assert_fs::prelude::*;
use common::{find_service_endpoint, setup_cluster};
use fs_extra::dir::CopyOptions;
use serial_test::serial;
use std::{
    path::{Path, PathBuf},
    process::Command,
};

static AME_FILE_NAME: &str = "ame.yaml";
static TARGET_NAMESPACE: &str = "ame-system";
static AME_SERVICE_NAME: &str = "ame-server-service";

async fn test_setup() -> Result<(), Box<dyn std::error::Error>> {
    std::env::set_var(
        "AME_ENDPOINT",
        find_service_endpoint(TARGET_NAMESPACE, AME_SERVICE_NAME).await?,
    );

    Ok(())
}

fn prepare_test_project(path_from_root: &str) -> Result<PathBuf, Box<dyn std::error::Error>> {
    let mut test_project_path = PathBuf::from(path_from_root);

    for i in 0..10 {
        if test_project_path.is_dir() && i < 9 {
            break;
        }

        if i < 9 {
            test_project_path = Path::new("..").join(test_project_path);
            continue;
        }

        return Err(format!(
            "failed to find test project directory: {}",
            test_project_path.display()
        ))?;
    }

    let temp_dir = assert_fs::TempDir::new()?.into_persistent();

    fs_extra::copy_items(
        &[test_project_path],
        temp_dir.path(),
        &CopyOptions::default(),
    )?;

    Ok(temp_dir
        .path()
        .join(Path::new(path_from_root).file_name().unwrap()))
}

#[test]
fn ame_file_already_exists() -> Result<(), Box<dyn std::error::Error>> {
    let temp = assert_fs::TempDir::new()?;
    let test_file = temp.child(AME_FILE_NAME);
    test_file.touch()?;

    let mut cmd = Command::cargo_bin("cli")?;

    cmd.current_dir(temp.path())
        .arg("init")
        .arg("myproject")
        .assert()
        .success();

    // If a file already exists we expect the CLI to inform the user and exist gracefully.
    // Therefore nothing should be written.
    test_file.assert("");

    Ok(())
}

#[test]
fn ame_file_does_not_exist() -> Result<(), Box<dyn std::error::Error>> {
    let temp = assert_fs::TempDir::new()?;
    let test_file = temp.child(AME_FILE_NAME);
    let mut cmd = Command::cargo_bin("cli")?;
    let project_id = "myproject";

    cmd.current_dir(temp.path())
        .arg("init")
        .arg(project_id)
        .assert()
        .success();

    test_file.assert(format!("projectid: {}\n", &project_id));

    Ok(())
}

#[tokio::test]
async fn ame_run_task() -> Result<(), Box<dyn std::error::Error>> {
    setup_cluster("ame-system").await?;
    let temp = prepare_test_project("test_data/test_projects/new_echo")?;
    println!("test project {}", temp.display());
    let mut cmd = Command::cargo_bin("cli")?;
    let task_id = "echo";
    test_setup().await?;

    let res = cmd
        .current_dir(temp)
        .arg("run")
        .arg(task_id)
        .assert()
        .success();

    let mut settings = insta::Settings::clone_current();
    settings.add_filter("time=\".*\"", "timestamp=\"redacted\"");
    settings.add_filter(
        "created virtual environment .*",
        "created virtual environment \"redacted\"",
    );
    settings.add_filter("creator .*", "creator \"redacted\"");
    let _guard = settings.bind_to_scope();
    insta::assert_snapshot!(&String::from_utf8(res.get_output().stdout.clone())?);

    Ok(())
}

#[tokio::test]
#[serial]
async fn ame_setup_cli() -> Result<(), Box<dyn std::error::Error>> {
    let temp = assert_fs::TempDir::new()?;

    let temp_path = temp.to_str().unwrap();

    let service_endpoint = find_service_endpoint(TARGET_NAMESPACE, AME_SERVICE_NAME)
        .await
        .unwrap();

    temp_env::with_vars(
        vec![("AME_ENDPOINT", None), ("XDG_CONFIG_HOME", Some(temp_path))],
        || {
            Command::cargo_bin("cli")
                .unwrap()
                .current_dir(temp_path)
                .arg("setup")
                .arg(service_endpoint.clone())
                .assert()
                .success();
        },
    );

    Ok(())
}

#[tokio::test]
#[serial]
async fn fail_bad_server_endpoint() -> Result<(), Box<dyn std::error::Error>> {
    let temp = assert_fs::TempDir::new()?;

    let temp_path = temp.to_str().unwrap();

    let service_endpoint = "wrong_endpoint".to_string();

    temp_env::with_vars(
        vec![("AME_ENDPOINT", None), ("XDG_CONFIG_HOME", Some(temp_path))],
        || {
            Command::cargo_bin("cli")
                .unwrap()
                .current_dir(temp_path)
                .arg("setup")
                .arg(service_endpoint.clone())
                .assert()
                .failure();
        },
    );

    Ok(())
}
