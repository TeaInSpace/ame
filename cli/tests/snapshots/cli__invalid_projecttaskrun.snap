---
source: cli/tests/cli.rs
expression: "&String::from_utf8(res.get_output().stderr.clone())?"
---
Error: Could not parse ame.yaml :(

Caused by:
    name: invalid type: sequence, expected a string at line 1 column 7

